lib/data/external/GCP/GcpManagedUpload.js

```diff
@@ -0,0 +1,432 @@
+     * @param {string} params.Bucket - bucket name
+     * @param {string} params.MPU - mpu bucket name
+     * @param {string} params.Overflow - overflow bucket name
+     * @param {string} params.Key - object key
+     * @param {object} options - config setting for GcpManagedUpload object
+     * @param {number} options.partSize - set object chunk size
+     * @param {number} options.queueSize - set the number of concurrent upload
+     * @return {object} - return an GcpManagedUpload object
+     */
+    constructor(service, params, options = {}) {
+        this.service = service;
+        this.params = params;
+        this.mainBucket =
+            this.params.Bucket || this.service.config.mainBucket;
+        this.mpuBucket = this.params.MPU || this.service.config.mpuBucket;
+        this.overflowBucket =
+            this.params.Overflow || this.service.config.overflowBucket;
+
+        this.partSize = minimumAllowedPartSize;
+        this.queueSize = options.queueSize || 4;
+        this.validateBody();
+        this.setPartSize();
+        // multipart information
+        this.parts = {};
+        this.uploadedParts = 0;
+        this.activeParts = 0;
+        this.partBuffers = [];
+        this.partQueue = [];
+        this.partBufferLength = 0;
+        this.totalChunkedBytes = 0;
+        this.partNumber = 0;
+    }
+
+    /**
+     * validateBody - validate that body contains data to upload. If body is not
+     * of type stream, it must then be of either string or buffer. If string,
+     * convert to a Buffer type and split into chunks if body is large enough
+     * @return {undefined}
+     */
+    validateBody() {
+        this.body = this.params.Body;
+        if (!this.body) {
+            this.error = errors.InvalidArgument
+                .customizeDescription('Body parameter is required');
+        } else {
+            if (this.params.ContentLength) {
+                this.totalBytes = this.params.ContentLength;
+            } else {
+                if (this.body instanceof stream && this.body.path &&
+                    typeof this.body.path === 'string') {
+                    this.totalBytes = lstatSync(this.body.path).size;
+                } else {
+                    if (typeof this.body === 'string') {
+                        this.body = Buffer.from(this.body);
+                    }
+                    this.totalBytes = this.body.byteLength;
+                }
+            }
+            this.totalBytes = this.totalBytes || 0;
+        }
+    }
+
+    setPartSize() {
+        if (this.totalBytes) {
+            const newPartSize =
+                Math.ceil(this.totalBytes / maximumAllowedPartCount);
+            if (newPartSize > this.partSize) this.partSize = newPartSize;
+            this.totalParts = Math.ceil(this.totalBytes / this.partSize);
+        }
+        if (this.body instanceof Buffer && this.totalParts > 1) {
+            this.slicedParts = sliceFn.call(this.body, this.partSize);
```

Comment:
I'm not sure if we need to use the `call()` method because we are not using any values from the `this` object in the function.

Reply:
Will fix this.